File: 0091-decode-ways/0091-decode-ways.py
class Solution:
    def numDecodings(self, s: str) -> int:
        def dfs(i):
            if i in memo:
                return memo[i]
            if s[i]=='0':
                return 0
            res = dfs(i+1)
            if i+1<len(s) and int(s[i:i+2])<27:
                res += dfs(i+2)
            memo[i] = res
            return memo[i]
        memo = {len(s):1}
        dfs(0)
        return memo[0] if 0 in memo else 0


# class Solution:
#     def numDecodings(self, s: str) -> int:
#         def solve(ptr):
#             if ptr>=len(s)-1: return 1
#             if mp[ptr]!=0: return mp[ptr]

#             num = (ord(s[ptr])-48)*10 + ord(s[ptr+1])-48
            
#             res=0
#             if num<10 or (num>26 and num%10==0):
#                 mp[ptr]=0
#                 mp[0]=0
#                 return mp[ptr]
#             if num%10!=0: res+=solve(ptr+1)
#             if num<=26: res+=solve(ptr+2)
            
#             mp[ptr]=res
#             return mp[ptr]
            

#         if s[0]=='0': return 0
#         mp=[0]*len(s)
#         solve(0)
#         return mp[0]

File: 0091-decode-ways/0091-decode-ways.java
class Solution {
    public int numDecodings(String s) {
        int n = s.length();
        int[] dp = new int[s.length() + 1];
        dp[0] = 1;
        
        for(int i=1; i<=n; ++i){
            dp[i] = s.charAt(i-1)=='0' ? 0 : dp[i-1];
            dp[i] += i>1 && s.charAt(i-2) != '0' && Integer.parseInt(s.substring(i-2, i)) <27
                ? dp[i-2] : 0;
        }
        
        return dp[n];
    }
}

// // DP - memo
//
// class Solution {
//     int[] memo;
//     public int numDecodings(String s) {
//         memo = new int[s.length() + 1];
//         Arrays.fill(memo, -1);
//         return help(s.length(), s);
//     }
//     int help(int i, String s){
//         if(i==0)
//             return 1;
        
//         if(memo[i] != -1) return memo[i];
        
//         int ans = s.charAt(i-1) != '0' ? help(i-1, s) : 0;
//         ans += i>1 && s.charAt(i-2) != '0' && Integer.parseInt(s.substring(i-2, i)) < 27 ?
//             help(i-2, s) : 0;
        
//         memo[i] = ans;
//         return ans;
//     }
// }